code/datums/unit_test.dm

```diff
@@ -0,0 +1,70 @@
+/*
+
+Usage:
+Override /Run() to run your test code
+
+Call Fail() to fail the test (You should specify a reason)
+
+You may use /New() and /Destroy() for setup/teardown respectively
+
+You can use the run_loc_bottom_left and run_loc_top_right to get turfs for testing
+
+*/
+
+GLOBAL_DATUM(current_test, /datum/unit_test)
+GLOBAL_VAR_INIT(failed_any_test, FALSE)
+
+/datum/unit_test
+	//Bit of metadata for the future maybe
+	var/list/procs_tested
+	
+	//usable vars
+	var/turf/run_loc_bottom_left
+	var/turf/run_loc_top_right
+
+	//internal shit
+	var/succeeded = TRUE
+	var/list/fail_reasons
+
+/datum/unit_test/New()
+	run_loc_bottom_left = locate(1, 1, 1)
+	run_loc_top_right = locate(5, 5, 1)
+
+/datum/unit_test/Destroy()
+	//clear the test area
+	for(var/atom/movable/AM in block(run_loc_bottom_left, run_loc_top_right))
+		qdel(AM)
+	return ..()
+
+/datum/unit_test/proc/Run()
+	Fail("[type]/Run() called parent or not implemented")
+
+/datum/unit_test/proc/Fail(reason = "No reason")
+	succeeded = FALSE
+
+	if(!istext(reason))
+		reason = "FORMATTED: [reason != null ? reason : "NULL"]"
+
+	LAZYADD(fail_reasons, reason)
+
+/world/proc/RunUnitTests()
+	CHECK_TICK
+	//can't rely on world/Error if this is off so results may be factually wrong
+#ifdef DEBUG

```

Comment:
This shouldn't have checkticks if no one is there to be connected? Also I feel like this should be a subsystem rather than some global. Just make the entire subsystem definition only compiled in if a unit test flag is set.

Reply:
If we don't CHECK_TICK the world will get pissy about infinite loops

Reply:
you can disable that at compile time by setting world.loop_checks to false.

Reply:
I'm reluctant to do that because it means if someone fucks up and does something that creates an infinite loop during tests we'd have travis jobs running forever with no clear explanation. I'd like to keep that infinite loop runtime visible. 

Also it makes the test environment more dissimilar to the live one if we run it all on one tick